### PR TITLE
Istio run tests

### PR DESCRIPTION
**Pull request by hunchback**
Added `scripts/ci/run-istio-tests.sh`  which should be integrated into the CI. 

Depends on:
- https://github.com/skydive-project/skydive/pull/1260
- https://github.com/skydive-project/skydive/pull/1256

### CODE REVIEW

File: topology/probes/k8s/graph.go [hunk 1]
```diff
@@ -57,84 +51,86 @@ func newMetadata(ty, namespace, name string, details interface{}) graph.Metadata
 	return m
 }
 
-func addMetadata(g *graph.Graph, n *graph.Node, details interface{}) {
+func newMetadata(ty, namespace, name string, details interface{}) graph.Metadata {
```
Comment (nplanel) [inline topology/probes/k8s/graph.go]: useful ?

File: topology/probes/k8s/networkpolicy.go [hunk 2]
```diff
@@ -77,7 +77,7 @@ func (n *networkPolicyProbe) OnAdd(obj interface{}) {
 	if np, ok := obj.(*v1beta1.NetworkPolicy); ok {
 		logging.GetLogger().Debugf("Adding %s", dumpNetworkPolicy(np))
 		n.graph.Lock()
-		npNode := newNode(n.graph, networkPolicyUID(np), n.newMetadata(np))
+		npNode := NewNode(n.graph, networkPolicyUID(np), n.newMetadata(np))
```
Comment (nplanel) [inline topology/probes/k8s/networkpolicy.go]: so you want to externalize the your newNode(g,a,b) equivalent to n.graph.NewNode(a,b) 
to NewNode(n.graph,a,b)  equivalent to n.graph.NewNode(a,b)  ?

Do see the problem here ?
(same for addMetadata ...)

File: scripts/ci/run-istio-tests.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+helm_teardown() {
+	[ "$KEEP_RESOURCES" = "true" ] || . "$DIR/install-istio.sh" stop
+	[ "$KEEP_RESOURCES" = "true" ] || . "$DIR/install-minikube.sh" stop
+}
+
+. "$DIR/run-tests-utils.sh"
+network_setup
+istio_setup
+WITH_ISTIO=true
+TEST_PATTERN=Istio
```
Comment (nplanel) [inline scripts/ci/run-istio-tests.sh]: why not running k8s and istio in one single : run-k8s-tests

Comment (nplanel) [reply scripts/ci/run-istio-tests.sh]: please use only run-k8s-tests.sh

File: Makefile [hunk 4]
```diff
@@ -56,7 +56,8 @@ ifeq ($(VERBOSE), false)
   VERBOSE_TESTS_FLAGS:=
 endif
 ifeq ($(COVERAGE), true)
-  EXTRA_ARGS+= -test.coverprofile=../functionals.cover
+  TEST_COVERPROFILE?=../functionals.cover
```
Comment (nplanel) [inline Makefile]: please don't change that, we don't need that

File: topology/probes/k8s/graph.go (hunk 1 above)
Comment (hunchback) [reply topology/probes/k8s/graph.go]: Its' a different function

File: topology/probes/k8s/networkpolicy.go (hunk 2 above)
Comment (hunchback) [reply topology/probes/k8s/networkpolicy.go]: moved code back to k8s/probes.go ... the NewNode is the generic function also used by istio, newNode is implemented in each k8s and istio and sets the correct manager (either "k8s" or "istio")

File: scripts/ci/run-istio-tests.sh (hunk 3 above)
Comment (hunchback) [reply scripts/ci/run-istio-tests.sh]: Istio can also be used on systems w/o k8s,  (integrating with consul), for this reason it make sense to separate the tests.

File: Makefile (hunk 4 above)
Comment (hunchback) [reply Makefile]: Notice that the original `run-k8s-tests.sh` renamed the `functionals.cover` - rather I gave the option to set the wanted target name.

File: topology/probes/k8s/graph.go [hunk 5]
```diff
@@ -57,84 +53,123 @@ func newMetadata(ty, namespace, name string, details interface{}) graph.Metadata
...
+func addOwnershipLink(g *graph.Graph, parent, child *graph.Node) *graph.Edge {
+	return AddOwnershipLink(ManagerValue, g, parent, child)
 }
 
-func dumpGraphNode(n *graph.Node) string {
-	ty, _ := n.GetFieldString("Type")
-	namespace, _ := n.GetFieldString("Namespace")
-	name, _ := n.GetFieldString("Name")
-	return fmt.Sprintf("%s{Namespace: %s, Name: %s}", ty, namespace, name)
+func NewNode(g *graph.Graph, i graph.Identifier, m graph.Metadata) *graph.Node {
```
Comment (nplanel) [inline topology/probes/k8s/graph.go]: why "C binding" function ?

File: scripts/ci/run-helm-tests.sh [hunk 6]
```diff
@@ -2,30 +2,8 @@
 
 set -v
```
Comment (nplanel) [inline scripts/ci/run-helm-tests.sh]: add set -e

Comment (hunchback) [reply scripts/ci/run-helm-tests.sh]: ok

File: topology/probes/k8s/graph.go (hunk 5 above)
Comment (hunchback) [reply topology/probes/k8s/graph.go]: @nplanel please explain ...

File: topology/probes/istio/cluster.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+func (p *clusterProbe) newMetadata(name string) graph.Metadata {
+	return newMetadata("cluster", "", name, nil)
+}
+
+func (p *clusterProbe) addNode(name string) {
+	p.graph.Lock()
+	defer p.graph.Unlock()
+
+	k8s.NewNode(p.graph, graph.GenID(), p.newMetadata(name))
```
Comment (nplanel) [inline topology/probes/istio/cluster.go]: this it why it's make no sense to export NewNode
you just pollute the k8s namespace

You function didn't do anything different than
p.graph.NewNode(graph.GenID(), p.newMetadata(name))

So use the object form, don't pollute namespace, it's confusing during reading and debugging

That's what I named kind of  "C binding" .... meaning do you plan really do call the function from C ?

File: topology/probes/k8s/cluster.go [hunk 8]
```diff
@@ -81,7 +71,7 @@ func (p *clusterProbe) addNode(name string) {
 	p.graph.Lock()
 	defer p.graph.Unlock()
 
-	clusterNode := newNode(p.graph, graph.GenID(), p.newMetadata(name))
+	clusterNode := NewNode(p.graph, graph.GenID(), p.newMetadata(name))
```
Comment (nplanel) [inline topology/probes/k8s/cluster.go]: same here

File: topology/probes/k8s/container.go [hunk 9]
```diff
@@ -124,10 +124,10 @@ func (c *containerProbe) onContainerAdd(pod *v1.Pod, container *v1.Container) {
 	containerNode := c.graph.GetNode(uid)
 	if containerNode == nil {
 		logging.GetLogger().Debugf("Adding %s", dumpContainer(pod, container))
-		containerNode = newNode(c.graph, uid, c.newMetadata(pod, container))
+		containerNode = NewNode(c.graph, uid, c.newMetadata(pod, container))
```
Comment (nplanel) [inline topology/probes/k8s/container.go]: same

File: topology/probes/k8s/cronjob.go [hunk 10]
```diff
@@ -59,7 +59,7 @@ func (p *cronJobProbe) OnAdd(obj interface{}) {
 		p.graph.Lock()
 		defer p.graph.Unlock()
 
-		newNode(p.graph, cronJobUID(cj), p.newMetadata(cj))
+		NewNode(p.graph, cronJobUID(cj), p.newMetadata(cj))
```
Comment (nplanel) [inline topology/probes/k8s/cronjob.go]: same

File: topology/probes/k8s/daemonset.go [hunk 11]
```diff
@@ -72,7 +72,7 @@ func (p *dsProbe) OnUpdate(oldObj, newObj interface{}) {
 		defer p.graph.Unlock()
 
 		if nsNode := p.graph.GetNode(dsUID(ds)); nsNode != nil {
-			addMetadata(p.graph, nsNode, ds)
+			AddMetadata(p.graph, nsNode, ds)
```
Comment (nplanel) [inline topology/probes/k8s/daemonset.go]: same

File: topology/probes/k8s/daemonset.go [hunk 12]
```diff
@@ -61,7 +61,7 @@ func (p *dsProbe) OnAdd(obj interface{}) {
 		p.graph.Lock()
 		defer p.graph.Unlock()
 
-		newNode(p.graph, dsUID(ds), p.newMetadata(ds))
+		NewNode(p.graph, dsUID(ds), p.newMetadata(ds))
```
Comment (nplanel) [inline topology/probes/k8s/daemonset.go]: same

File: topology/probes/k8s/deployment.go [hunk 13]
```diff
@@ -63,7 +63,7 @@ func (p *deployProbe) OnAdd(obj interface{}) {
 		p.graph.Lock()
 		defer p.graph.Unlock()
 
-		newNode(p.graph, deployUID(deploy), p.newMetadata(deploy))
+		NewNode(p.graph, deployUID(deploy), p.newMetadata(deploy))
```
Comment (nplanel) [inline topology/probes/k8s/deployment.go]: and so on ....

File: topology/probes/k8s/networkpolicy.go (hunk 2 above)
Comment (nplanel) [reply topology/probes/k8s/networkpolicy.go]: in k8s probe

func NewNode(g *graph.Graph, i graph.Identifier, m graph.Metadata) *graph.Node {
	return g.NewNode(i, m, hostID)
}

where is the gain/enhancement ?
 1/ there are no value added
 2/ it's introduce confusion during rereading / debugging
 3/ you pollute the k8s namespace by exporting these functions

In this case : 
n.graph.NewNode(networkPolicyUID(np), n.newMetadata(np))

will do the same thing

Comment (hunchback) [reply topology/probes/k8s/networkpolicy.go]: setting the correct `hostID` is the reason for this function

File: topology/probes/istio/cluster.go (hunk 7 above)
Comment (hunchback) [reply topology/probes/istio/cluster.go]: as explained the reason is that I want to hide the correct setting of `hostID`

File: scripts/ci/run-istio-tests.sh [hunk 14]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+set -v
+set -e
+
+DIR="$(dirname "$0")"
+
+istio_setup() {
+        . "$DIR/install-minikube.sh" install
```
Comment (lebauce) [inline scripts/ci/run-istio-tests.sh]: nit: fix indentation

File: scripts/ci/run-tests-utils.sh [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  	    export TEST_COVERPROFILE=../functionals-$BACKEND.cover
+	fi
+
+        make test.functionals.batch \
+                GOFLAGS="$GOFLAGS" VERBOSE=true TAGS="$TAGS" GORACE="history_size=5" TIMEOUT=20m \
+		WITH_HELM="$WITH_HELM" WITH_EBPF="$WITH_EBPF" WITH_K8S="$WITH_K8S" WITH_ISTIO="$WITH_ISTIO" \
+                ARGS="$ARGS" TEST_PATTERN=$TEST_PATTERN 2>&1 | tee $LOGFILE
+
+	go get -f -u github.com/tebeka/go2xunit
+        go2xunit -fail -fail-on-race -suite-name-prefix tests \
```
Comment (lebauce) [inline scripts/ci/run-tests-utils.sh]: nit: fix indentation

File: topology/probes/k8s/networkpolicy.go (hunk 2 above)
Comment (lebauce) [reply topology/probes/k8s/networkpolicy.go]: Thinking out loud : now that we have an "Origin" field and with the PR (https://github.com/skydive-project/skydive/pull/1210), maybe we could set the `host` attribute of the analyzer graph to the empty string.
This way, you won't have to explicitely specify an empty host.

File: scripts/ci/run-istio-tests.sh (hunk 14 above)
Comment (hunchback) [reply scripts/ci/run-istio-tests.sh]: done

File: scripts/ci/run-tests-utils.sh (hunk 15 above)
Comment (hunchback) [reply scripts/ci/run-tests-utils.sh]: done

File: topology/probes/k8s/networkpolicy.go (hunk 2 above)
Comment (hunchback) [reply topology/probes/k8s/networkpolicy.go]: @lebauce / @nplanel - I am not changing the current implementation - only making the existing functions public (and adding the manager field). I am fine with any such change but would like this to be done as a separate PR.

**Pull request closed without merging**